libp2p/connmanager.nim

```diff
@@ -0,0 +1,276 @@
+    return c.muxed[conn].muxer
+
+proc storeConn*(c: ConnManager, conn: Connection) =
+  ## store a connection
+  ##
+
+  if isNil(conn):
+    raise newException(CatchableError, "connection cannot be nil")
+
+  if isNil(conn.peerInfo):
+    raise newException(CatchableError, "empty peer info")
+
+  let peerInfo = conn.peerInfo
+  if c.conns.getOrDefault(peerInfo.peerId).len > c.maxConns:
+    trace "too many connections", peer = $conn.peerInfo,
+                                  conns = c.conns
+                                  .getOrDefault(peerInfo.peerId).len
+
+    raise newTooManyConnections()
+
+  if peerInfo.peerId notin c.conns:
+    c.conns[peerInfo.peerId] = initHashSet[Connection]()
+
+  c.conns[peerInfo.peerId].incl(conn)
+
+  # launch on close listener
+  asyncCheck c.onClose(conn)
+  libp2p_peers.set(c.conns.len.int64)
+
+proc storeOutgoing*(c: ConnManager, conn: Connection) =
+  conn.dir = Direction.Out
+  c.storeConn(conn)
+
+proc storeIncoming*(c: ConnManager, conn: Connection) =
+  conn.dir = Direction.In
+  c.storeConn(conn)
+
+proc storeMuxer*(c: ConnManager,
+                 muxer: Muxer,
+                 handle: Future[void] = nil) =
+  ## store the connection and muxer
+  ##
+
+  if isNil(muxer):
+    raise newException(CatchableError, "muxer cannot be nil")
+
+  if isNil(muxer.connection):
+    raise newException(CatchableError, "muxer's connection cannot be nil")
+
+  c.muxed[muxer.connection] = MuxerHolder(
+    muxer: muxer,
+    handle: handle)
+
+  trace "storred connection", connections = c.conns.len
+
+proc getMuxedStream*(c: ConnManager,
+                     peerInfo: PeerInfo,
+                     dir: Direction): Future[Connection] {.async, gcsafe.} =
+  ## get a muxed stream for the provided peer
+  ## with the given direction
+  ##
+
+  let muxer = c.selectMuxer(c.selectConn(peerInfo, dir)) # always get the first muxer here
+  if not(isNil(muxer)):
+    return await muxer.newStream()
+
+proc getMuxedStream*(c: ConnManager,
+                     peerInfo: PeerInfo): Future[Connection] {.async, gcsafe.} =
+  ## get a muxed stream for the passed peer from any connection
+  ##
+
+  let muxer = c.selectMuxer(c.selectConn(peerInfo)) # always get the first muxer here
+  if not(isNil(muxer)):
+    return await muxer.newStream()
+
+proc getMuxedStream*(c: ConnManager,
+                     conn: Connection): Future[Connection] {.async, gcsafe.} =
+  ## get a muxed stream for the passed connection
+  ##
+
+  let muxer = c.selectMuxer(conn) # always get the first muxer here
+  if not(isNil(muxer)):
+    return await muxer.newStream()
+
+proc dropPeer*(c: ConnManager, peerInfo: PeerInfo) {.async.} =
+  ## drop connections and cleanup resources for peer
+  ##
+
+  for conn in c.conns.getOrDefault(peerInfo.peerId):
+    if not(isNil(conn)):
+      await c.cleanupConn(conn)
+
+proc close*(c: ConnManager) {.async.} =
+  ## cleanup resources for the connection
+  ## manager
+  ##
+
+  for conns in toSeq(c.conns.values):
```

Comment:
I guess copying mitigates races, but still not convinced 100%, maybe interlocking is better

Reply:
Usually copying is fine here, but since this is close, we might want to lock. I would avoid it for now tho.